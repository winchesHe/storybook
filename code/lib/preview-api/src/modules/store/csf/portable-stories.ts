/* eslint-disable @typescript-eslint/naming-convention */
import { isExportStory } from '@storybook/csf';
import type {
  Renderer,
  Args,
  ComponentAnnotations,
  LegacyStoryAnnotationsOrFn,
  ProjectAnnotations,
  ComposedStoryPlayFn,
  ComposeStoryFn,
  Store_CSFExports,
  StoryContext,
  Parameters,
  ComposedStoryFn,
  StrictArgTypes,
  PlayFunctionContext,
} from '@storybook/types';

import { HooksContext } from '../../../addons';
import { composeConfigs } from './composeConfigs';
import { prepareContext, prepareStory } from './prepareStory';
import { normalizeStory } from './normalizeStory';
import { normalizeComponentAnnotations } from './normalizeComponentAnnotations';
import { getValuesFromArgTypes } from './getValuesFromArgTypes';
import { normalizeProjectAnnotations } from './normalizeProjectAnnotations';

let GLOBAL_STORYBOOK_PROJECT_ANNOTATIONS = composeConfigs([]);

export function getPortableStoryWrapperId(storyId: string) {
  return `storybook-story-${storyId}`;
}

export function setProjectAnnotations<TRenderer extends Renderer = Renderer>(
  projectAnnotations: ProjectAnnotations<TRenderer> | ProjectAnnotations<TRenderer>[]
) {
  const annotations = Array.isArray(projectAnnotations) ? projectAnnotations : [projectAnnotations];
  GLOBAL_STORYBOOK_PROJECT_ANNOTATIONS = composeConfigs(annotations);
}

export function composeStory<TRenderer extends Renderer = Renderer, TArgs extends Args = Args>(
  storyAnnotations: LegacyStoryAnnotationsOrFn<TRenderer>,
  componentAnnotations: ComponentAnnotations<TRenderer, TArgs>,
  projectAnnotations: ProjectAnnotations<TRenderer> = GLOBAL_STORYBOOK_PROJECT_ANNOTATIONS as ProjectAnnotations<TRenderer>,
  defaultConfig: ProjectAnnotations<TRenderer> = {},
  exportsName?: string
): ComposedStoryFn<TRenderer, Partial<TArgs>> {
  if (storyAnnotations === undefined) {
    throw new Error('Expected a story but received undefined.');
  }

  // users might pass an empty object instead of undefined e.g. composeStory(story, meta, {}, exportsName)
  // and likely they expect the default project annotations to be used instead of completely resetting them
  if (typeof projectAnnotations === 'object' && Object.keys(projectAnnotations).length === 0) {
    projectAnnotations = GLOBAL_STORYBOOK_PROJECT_ANNOTATIONS as ProjectAnnotations<TRenderer>;
  }

  // @TODO: Support auto title

  componentAnnotations.title = componentAnnotations.title ?? 'ComposedStory';
  const normalizedComponentAnnotations =
    normalizeComponentAnnotations<TRenderer>(componentAnnotations);

  const storyName =
    exportsName ||
    storyAnnotations.storyName ||
    storyAnnotations.story?.name ||
    storyAnnotations.name ||
    'Unnamed Story';

  const normalizedStory = normalizeStory<TRenderer>(
    storyName,
    storyAnnotations,
    normalizedComponentAnnotations
  );

  const normalizedProjectAnnotations = normalizeProjectAnnotations<TRenderer>(
    composeConfigs([projectAnnotations, defaultConfig])
  );

  const story = prepareStory<TRenderer>(
    normalizedStory,
    normalizedComponentAnnotations,
    normalizedProjectAnnotations
  );

  const defaultGlobals = getValuesFromArgTypes(projectAnnotations.globalTypes);

  const context: StoryContext<TRenderer> = {
    hooks: new HooksContext(),
    globals: {
      ...defaultGlobals,
      ...projectAnnotations.globals,
    },
    args: { ...story.initialArgs },
    viewMode: 'story',
    loaded: {},
    abortSignal: null as unknown as AbortSignal,
    canvasElement: null,
    ...story,
  };

  const playFunction = story.playFunction
    ? async (extraContext: Partial<PlayFunctionContext<TRenderer, TArgs>>) =>
        story.playFunction!({
          ...context,
          ...extraContext,
          // if canvasElement is not provided, we default to the root element, which comes from a decorator
          // the decorator has to be implemented in the defaultAnnotations of each integrator of portable stories
          canvasElement:
            extraContext?.canvasElement ??
            globalThis.document?.getElementById(getPortableStoryWrapperId(context.id)),
        })
    : undefined;

  const composedStory: ComposedStoryFn<TRenderer, Partial<TArgs>> = Object.assign(
    function storyFn(extraArgs?: Partial<TArgs>) {
      context.args = {
        ...context.initialArgs,
        ...extraArgs,
      };

      return story.unboundStoryFn(prepareContext(context));
    },
    {
      id: story.id,
      storyName,
      args: story.initialArgs as Partial<TArgs>,
      parameters: story.parameters as Parameters,
      argTypes: story.argTypes as StrictArgTypes<TArgs>,
      play: playFunction as ComposedStoryPlayFn<TRenderer, TArgs> | undefined,
    }
  );

  return composedStory;
}

export function composeStories<TModule extends Store_CSFExports>(
  storiesImport: TModule,
  globalConfig: ProjectAnnotations<Renderer>,
  composeStoryFn: ComposeStoryFn
) {
  const { default: meta, __esModule, __namedExportsOrder, ...stories } = storiesImport;
  const composedStories = Object.entries(stories).reduce((storiesMap, [exportsName, story]) => {
    if (!isExportStory(exportsName, meta)) {
      return storiesMap;
    }

    const result = Object.assign(storiesMap, {
      [exportsName]: composeStoryFn(
        story as LegacyStoryAnnotationsOrFn,
        meta,
        globalConfig,
        exportsName
      ),
    });
    return result;
  }, {});

  return composedStories;
}
