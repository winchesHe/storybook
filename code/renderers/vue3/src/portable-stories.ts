import {
  composeStory as originalComposeStory,
  composeStories as originalComposeStories,
  setProjectAnnotations as originalSetProjectAnnotations,
  getPortableStoryWrapperId,
} from '@storybook/preview-api';
import type {
  Args,
  ProjectAnnotations,
  StoryAnnotationsOrFn,
  Store_CSFExports,
  StoriesWithPartialProps,
} from '@storybook/types';

import * as vueProjectAnnotations from './entry-preview';
import type { Meta } from './public-types';
import type { VueRenderer } from './types';

const defaultProjectAnnotations: ProjectAnnotations<VueRenderer> = {
  ...vueProjectAnnotations,
  decorators: [
    function addStorybookId(story, { id }) {
      return {
        components: { story },
        template: `<div data-story="true" id="${getPortableStoryWrapperId(id)}">
          <story />
        </div>`,
      };
    },
  ],
};

/** Function that sets the globalConfig of your Storybook. The global config is the preview module of your .storybook folder.
 *
 * It should be run a single time, so that your global config (e.g. decorators) is applied to your stories when using `composeStories` or `composeStory`.
 *
 * Example:
 *```jsx
 * // setup.js (for jest)
 * import { setProjectAnnotations } from '@storybook/vue3';
 * import projectAnnotations from './.storybook/preview';
 *
 * setProjectAnnotations(projectAnnotations);
 *```
 *
 * @param projectAnnotations - e.g. (import projectAnnotations from '../.storybook/preview')
 */
export function setProjectAnnotations(
  projectAnnotations: ProjectAnnotations<VueRenderer> | ProjectAnnotations<VueRenderer>[]
) {
  originalSetProjectAnnotations<VueRenderer>(projectAnnotations);
}

/**
 * Function that will receive a story along with meta (e.g. a default export from a .stories file)
 * and optionally projectAnnotations e.g. (import * from '../.storybook/preview)
 * and will return a composed component that has all args/parameters/decorators/etc combined and applied to it.
 *
 *
 * It's very useful for reusing a story in scenarios outside of Storybook like unit testing.
 *
 * Example:
 *```jsx
 * import { render } from '@testing-library/vue';
 * import { composeStory } from '@storybook/vue3';
 * import Meta, { Primary as PrimaryStory } from './Button.stories';
 *
 * const Primary = composeStory(PrimaryStory, Meta);
 *
 * test('renders primary button with Hello World', () => {
 *   const { getByText } = render(Primary({label: "Hello world"}));
 *   expect(getByText(/Hello world/i)).not.toBeNull();
 * });
 *```
 *
 * @param story
 * @param componentAnnotations - e.g. (import Meta from './Button.stories')
 * @param [projectAnnotations] - e.g. (import * as projectAnnotations from '../.storybook/preview') this can be applied automatically if you use `setProjectAnnotations` in your setup files.
 * @param [exportsName] - in case your story does not contain a name and you want it to have a name.
 */
export function composeStory<TArgs extends Args = Args>(
  story: StoryAnnotationsOrFn<VueRenderer, TArgs>,
  componentAnnotations: Meta<TArgs | any>,
  projectAnnotations?: ProjectAnnotations<VueRenderer>,
  exportsName?: string
) {
  return originalComposeStory<VueRenderer, TArgs>(
    story as StoryAnnotationsOrFn<VueRenderer, Args>,
    componentAnnotations,
    projectAnnotations,
    defaultProjectAnnotations,
    exportsName
  );
}

/**
 * Function that will receive a stories import (e.g. `import * as stories from './Button.stories'`)
 * and optionally projectAnnotations (e.g. `import * from '../.storybook/preview`)
 * and will return an object containing all the stories passed, but now as a composed component that has all args/parameters/decorators/etc combined and applied to it.
 *
 *
 * It's very useful for reusing stories in scenarios outside of Storybook like unit testing.
 *
 * Example:
 *```jsx
 * import { render } from '@testing-library/vue';
 * import { composeStories } from '@storybook/vue3';
 * import * as stories from './Button.stories';
 *
 * const { Primary, Secondary } = composeStories(stories);
 *
 * test('renders primary button with Hello World', () => {
 *   const { getByText } = render(Primary({label: "Hello world"}));
 *   expect(getByText(/Hello world/i)).not.toBeNull();
 * });
 *```
 *
 * @param csfExports - e.g. (import * as stories from './Button.stories')
 * @param [projectAnnotations] - e.g. (import * as projectAnnotations from '../.storybook/preview') this can be applied automatically if you use `setProjectAnnotations` in your setup files.
 */
export function composeStories<TModule extends Store_CSFExports<VueRenderer, any>>(
  csfExports: TModule,
  projectAnnotations?: ProjectAnnotations<VueRenderer>
) {
  // @ts-expect-error Deep down TRenderer['canvasElement'] resolves to canvasElement: unknown but VueRenderer uses WebRenderer where canvasElement is HTMLElement, so the types clash
  const composedStories = originalComposeStories(csfExports, projectAnnotations, composeStory);

  return composedStories as unknown as Omit<
    StoriesWithPartialProps<VueRenderer, TModule>,
    keyof Store_CSFExports
  >;
}
