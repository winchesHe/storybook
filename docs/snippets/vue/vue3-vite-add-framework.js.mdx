```js
// .storybook/main.js
export default {
  // ...
  framework: '@storybook/vue3-vite', // ðŸ‘ˆ Add this
};
```
