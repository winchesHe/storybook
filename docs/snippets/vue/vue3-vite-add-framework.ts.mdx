```ts
// .storybook/main.ts
import { StorybookConfig } from '@storybook/nextjs';

const config: StorybookConfig = {
  // ...
  framework: '@storybook/vue3-vite', // ðŸ‘ˆ Add this
};

export default config;
```
