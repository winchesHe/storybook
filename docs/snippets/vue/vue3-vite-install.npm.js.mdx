```shell
npm install --save-dev @storybook/vue3-vite
```
