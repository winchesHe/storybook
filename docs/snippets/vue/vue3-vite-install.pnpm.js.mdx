```shell
pnpm install --save-dev @storybook/vue3-vite
```
