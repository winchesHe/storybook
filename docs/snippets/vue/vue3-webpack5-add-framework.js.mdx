```js
// .storybook/main.js
export default {
  // ...
  framework: '@storybook/vue3-webpack5', // ðŸ‘ˆ Add this
};
```
