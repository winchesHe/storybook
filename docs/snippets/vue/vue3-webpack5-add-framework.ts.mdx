```ts
// .storybook/main.ts
import { StorybookConfig } from '@storybook/vue3-webpack5';

const config: StorybookConfig = {
  // ...
  framework: '@storybook/vue3-webpack5', // ðŸ‘ˆ Add this
};

export default config;
```
