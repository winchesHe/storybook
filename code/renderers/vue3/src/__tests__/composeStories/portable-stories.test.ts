/// <reference types="@testing-library/jest-dom" />;
import { it, expect, vi, describe } from 'vitest';
import { render, screen } from '@testing-library/vue';
import { addons } from '@storybook/preview-api';
import { expectTypeOf } from 'expect-type';
import type { Meta } from '@storybook/vue3';

import * as stories from './Button.stories';
import type Button from './Button.vue';
import { composeStories, composeStory, setProjectAnnotations } from '../../portable-stories';

// example with composeStories, returns an object with all stories composed with args/decorators
const { CSF3Primary } = composeStories(stories);

// example with composeStory, returns a single story composed with args/decorators
const Secondary = composeStory(stories.CSF2Secondary, stories.default);

it('renders primary button', () => {
  render(CSF3Primary({ label: 'Hello world' }));
  const buttonElement = screen.getByText(/Hello world/i);
  expect(buttonElement).toBeInTheDocument();
});

it('reuses args from composed story', () => {
  render(Secondary());
  const buttonElement = screen.getByRole('button');
  expect(buttonElement.textContent).toEqual(Secondary.args.label);
});

it('myClickEvent handler is called', async () => {
  const myClickEventSpy = vi.fn();
  render(Secondary({ onMyClickEvent: myClickEventSpy }));
  const buttonElement = screen.getByRole('button');
  buttonElement.click();
  expect(myClickEventSpy).toHaveBeenCalled();
});

it('reuses args from composeStories', () => {
  const { getByText } = render(CSF3Primary());
  const buttonElement = getByText(/foo/i);
  expect(buttonElement).toBeInTheDocument();
});

describe('projectAnnotations', () => {
  it('renders with default projectAnnotations', () => {
    const WithEnglishText = composeStory(stories.CSF2StoryWithLocale, stories.default);
    const { getByText } = render(WithEnglishText());
    const buttonElement = getByText('Hello!');
    expect(buttonElement).toBeInTheDocument();
  });

  it('renders with custom projectAnnotations via composeStory params', () => {
    const WithPortugueseText = composeStory(stories.CSF2StoryWithLocale, stories.default, {
      globalTypes: { locale: { defaultValue: 'pt' } } as any,
    });
    const { getByText } = render(WithPortugueseText());
    const buttonElement = getByText('OlÃ¡!');
    expect(buttonElement).toBeInTheDocument();
  });

  it('renders with custom projectAnnotations via setProjectAnnotations', () => {
    setProjectAnnotations([{ parameters: { injected: true } }]);
    const Story = composeStory(stories.CSF2StoryWithLocale, stories.default);
    expect(Story.parameters?.injected).toBe(true);
  });
});

describe('CSF3', () => {
  it('renders with inferred globalRender', () => {
    const Primary = composeStory(stories.CSF3Button, stories.default);

    render(Primary({ label: 'Hello world' }));
    const buttonElement = screen.getByText(/Hello world/i);
    expect(buttonElement).toBeInTheDocument();
  });

  it('renders with custom render function', () => {
    const Primary = composeStory(stories.CSF3ButtonWithRender, stories.default);

    render(Primary());
    expect(screen.getByTestId('custom-render')).toBeInTheDocument();
  });

  it('renders with play function', async () => {
    const CSF3InputFieldFilled = composeStory(stories.CSF3InputFieldFilled, stories.default);

    render(CSF3InputFieldFilled());

    await CSF3InputFieldFilled.play!();

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');
  });
});

// common in addons that need to communicate between manager and preview
it('should pass with decorators that need addons channel', () => {
  const PrimaryWithChannels = composeStory(stories.CSF3Primary, stories.default, {
    decorators: [
      (StoryFn: any) => {
        addons.getChannel();
        return StoryFn();
      },
    ],
  });
  render(PrimaryWithChannels({ label: 'Hello world' }));
  const buttonElement = screen.getByText(/Hello world/i);
  expect(buttonElement).not.toBeNull();
});

describe('ComposeStories types', () => {
  it('Should support typescript operators', () => {
    type ComposeStoriesParam = Parameters<typeof composeStories>[0];

    expectTypeOf({
      ...stories,
      default: stories.default as Meta<typeof Button>,
    }).toMatchTypeOf<ComposeStoriesParam>();

    expectTypeOf({
      ...stories,
      default: stories.default satisfies Meta<typeof Button>,
    }).toMatchTypeOf<ComposeStoriesParam>();
  });
});

// Batch snapshot testing
const testCases = Object.values(composeStories(stories)).map((Story) => [Story.storyName, Story]);
it.each(testCases)('Renders %s story', async (_storyName, Story) => {
  if (typeof Story === 'string' || _storyName === 'CSF2StoryWithParamsAndDecorator') {
    return;
  }

  await new Promise((resolve) => setTimeout(resolve, 0));

  const tree = await render(Story());
  expect(tree.baseElement).toMatchSnapshot();
});
