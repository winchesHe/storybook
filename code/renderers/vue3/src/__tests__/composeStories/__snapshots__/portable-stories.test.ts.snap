// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders CSF2Secondary story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-2-secondary"
    >
      <button
        class="storybook-button storybook-button--secondary storybook-button--medium"
        type="button"
      >
        label coming from story args!
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders CSF2StoryWithLocale story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-2-story-with-locale"
    >
      <div>
        <p>
          locale: undefined
        </p>
        <button
          class="storybook-button storybook-button--secondary storybook-button--medium"
          type="button"
        >
          Hello!
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Renders CSF3Button story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-3-button"
    >
      <button
        class="storybook-button storybook-button--secondary storybook-button--medium"
        type="button"
      >
        foo
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders CSF3ButtonWithRender story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-3-button-with-render"
    >
      <div>
        <p
          data-testid="custom-render"
        >
          I am a custom render function
        </p>
        <button
          class="storybook-button storybook-button--secondary storybook-button--medium"
          type="button"
        >
          foo
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Renders CSF3InputFieldFilled story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-3-input-field-filled"
    >
      <input
        data-testid="input"
      />
    </div>
  </div>
</body>
`;

exports[`Renders CSF3Primary story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-3-primary"
    >
      <button
        class="storybook-button storybook-button--primary storybook-button--large"
        type="button"
      >
        foo
      </button>
    </div>
  </div>
</body>
`;
